pkg/config/example_conversions.go

```diff
@@ -0,0 +1,139 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8sschema "k8s.io/apimachinery/pkg/runtime/schema"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/yaml"
+
+	"github.com/crossplane/upjet/pkg/config/conversion"
+)
+
+func ConvertSingletonListToEmbeddedObject(pc *Provider, startPath string) error { //nolint:gocyclo
+	resourceRegistry := prepareResourceRegistry(pc)
+	err := filepath.Walk(startPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrap(err, "walk failed")
+		}
+
+		var convertedFileContent string
+		if !info.IsDir() && strings.HasSuffix(info.Name(), ".yaml") {
+			log.Printf("Converting: %s\n", path)
+			content, err := os.ReadFile(path) //nolint:gosec
+			if err != nil {
+				return errors.Wrap(err, "failed to read file")
+			}
+
+			examples, err := decodeExamples(string(content))
+			if err != nil {
+				return errors.Wrap(err, "failed to decode examples")
+			}
+
+			rootResource := resourceRegistry[fmt.Sprintf("%s/%s", examples[0].GroupVersionKind().Kind, examples[0].GroupVersionKind().Group)]
+			if rootResource == nil {
+				return nil
+			}
+
+			newPath := strings.Replace(path, examples[0].GroupVersionKind().Version, rootResource.Version, -1) //nolint:gocritic
+			if path == newPath {
+				return nil
+			}
+			annotationValue := strings.ToLower(fmt.Sprintf("%s/%s/%s", rootResource.ShortGroup, rootResource.Version, rootResource.Kind))
+			for _, e := range examples {
+				if resource, ok := resourceRegistry[fmt.Sprintf("%s/%s", e.GroupVersionKind().Kind, e.GroupVersionKind().Group)]; ok {
+					conversionPaths := resource.CRDListConversionPaths()
+					if conversionPaths != nil && e.GroupVersionKind().Version != resource.Version {
+						for i, cp := range conversionPaths {
+							conversionPaths[i] = "spec.forProvider." + cp
+						}
+						converted, err := conversion.Convert(e.Object, conversionPaths, conversion.ToEmbeddedObject)
+						if err != nil {
+							return errors.Wrap(err, "failed to convert example to embedded object")
+						}
+						e.Object = converted
+						e.SetGroupVersionKind(k8sschema.GroupVersionKind{
+							Group:   e.GroupVersionKind().Group,
+							Version: resource.Version,
+							Kind:    e.GetKind(),
+						})
+					}
+					annotations := e.GetAnnotations()
+					annotations["meta.upbound.io/example-id"] = annotationValue
+					e.SetAnnotations(annotations)
+				}
+			}
+			convertedFileContent = "# SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>\n#\n# SPDX-License-Identifier: CC0-1.0\n\n"
```

Comment:
We had better load such header from a file (like the `hack/boilerplate.go.txt` files that are common in the official provider repositories) since we are positioning this as a library that the provider authors can use.

Reply:
Added a parameter to method for specifying the license header file path.